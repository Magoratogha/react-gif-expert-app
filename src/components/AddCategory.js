import { useState } from "react";
import PropTypes from 'prop-types';

function AddCategory({ submitCallbackFn }) {

    const [inputValue, setInputValue] = useState('');

    function handleInputChange (event) {
        setInputValue(event.target.value);
    };

    function handleInputSubmit(event) {
        event.preventDefault();
        if (inputValue.trim().length > 2) {
            submitCallbackFn((previousValue) => [inputValue, ...previousValue]);
            setInputValue('');
        }
    };

    return (
        <form onSubmit={ handleInputSubmit }>
            <input
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }    
            />
        </form>
    );
}

AddCategory.propTypes = {
    submitCallbackFn: PropTypes.func.isRequired
}

export default AddCategory;