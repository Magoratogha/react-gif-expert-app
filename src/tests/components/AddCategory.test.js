import { render, queryByAttribute } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AddCategory from "../../components/AddCategory";

describe('Test in <AddCategory />', () => {
    
    const callbackFnMock = jest.fn();
    const user = userEvent.setup();
    
    test('should render properly', () => {
        const { container } = render(<AddCategory submitCallbackFn={ callbackFnMock } />);
        expect(container).toMatchSnapshot();
    });

    test('should change the input value when user type a new input', async() => {
        const { container } = render(<AddCategory submitCallbackFn={ callbackFnMock } />);
        const value = 'This is a test';
        const inputElement = queryByAttribute('type', container, 'text');
        await user.type(inputElement, value);
        expect(inputElement).toHaveValue(value);
    });

    test('should not call submitCallbackFn if user type a input with length < 3 characters', async() => {
        const { container } = render(<AddCategory submitCallbackFn={ callbackFnMock } />);
        const value = 'Th';
        const inputElement = queryByAttribute('type', container, 'text');
        await user.type(inputElement, value);
        await user.type(inputElement, '{Enter}');
        expect(callbackFnMock).not.toHaveBeenCalled();
    });

    test('should call submitCallbackFn if user type a valid input and then clean the value', async() => {
        const { container } = render(<AddCategory submitCallbackFn={ callbackFnMock } />);
        const value = 'Test value';
        const inputElement = queryByAttribute('type', container, 'text');
        await user.type(inputElement, value);
        await user.type(inputElement, '{Enter}');
        expect(callbackFnMock).toHaveBeenCalledWith(expect.any(Function));
        expect(inputElement).toHaveValue('');
    });
});