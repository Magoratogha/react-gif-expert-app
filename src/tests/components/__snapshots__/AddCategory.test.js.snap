// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <AddCategory /> should render properly 1`] = `
<div>
  <form>
    <input
      type="text"
      value=""
    />
  </form>
</div>
`;
